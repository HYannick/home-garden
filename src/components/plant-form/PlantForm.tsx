/** @jsx jsx */
import React from 'react';
import { Field, Formik } from 'formik';
import { css, jsx } from '@emotion/core';
import { Trans, useTranslation } from 'react-i18next';
import plantCreateValidation from '../../pages/create/plant-create.validation';
import InputField from '../forms/InputField';
import DatePicker from '../forms/DatePicker';
import Switch from '../forms/Switch';
import MoistureSensorInput from '../forms/MoistureSensorField';
import InputRange from '../forms/InputRange';
import { Button } from '../Button';
import { FormCreate, ImageCreate } from './PlantForm.styled';

interface PlantFormProps {
  onSubmit: any,
  initialValues: any,
  submitLabel: string
}

const PlantForm: React.FC<PlantFormProps> = ({ onSubmit, initialValues, submitLabel }) => {
  const { t } = useTranslation();
  return (
    <Formik
      validate={(values) => plantCreateValidation(values, t)}
      onSubmit={onSubmit}
      initialValues={initialValues}
      render={(props) => {
        const { errors, touched, isSubmitting, values, setFieldValue } = props;
        const resetField = (field: string) => setFieldValue(field, '');
        return (
          <FormCreate>
            <Field
              type="text"
              name="picture"
              component={ImageCreate}
              onImageLoaded={(imgData: { src: string, url: string }) => setFieldValue('picture', imgData.src)}/>
            <FormCreate.Wrapper>
              <FormCreate.Title>
                <p>Plant infos</p>
              </FormCreate.Title>
              <div css={css`flex:1; padding: 0 2.5rem 0 1rem; margin-top: 2rem;`}>
                <FormCreate.Control>
                  <Field type="text" name="custom_name" component={InputField} label={t('plant_form.field_name')}/>
                </FormCreate.Control>
                <FormCreate.Control>
                  <Field
                    type="date"
                    name="last_watering_date"
                    component={DatePicker}
                    label={t('plant_form.field_last_watering_date')}
                    onDateSelected={(date: Date) => setFieldValue('last_watering_date', date)}/>
                  {errors.last_watering_date && touched.last_watering_date &&
                  <FormCreate.ErrorField>{errors.last_watering_date}</FormCreate.ErrorField>}
                </FormCreate.Control>
                <FormCreate.Control>
                  <Field
                    type="checkbox"
                    name="has_moisture_sensor"
                    component={Switch}
                    label={t('plant_form.field_has_moisture_sensor')}
                    onChange={(has_moisture_sensor: boolean) => {
                      if (!has_moisture_sensor) resetField('sensor_id');
                      setFieldValue('has_moisture_sensor', has_moisture_sensor);
                    }}
                  />
                  {values.has_moisture_sensor && (
                    <MoistureSensorInput
                      testId="moisture-sensor-input"
                      {...props} resetField={resetField}
                      t={t}/>
                  )}
                </FormCreate.Control>
                {
                  !values.has_moisture_sensor && (
                    <FormCreate.Control data-testid="watering-frequency-input">
                      <Field
                        type="range"
                        name="watering_frequency"
                        label={t('plant_form.field_watering_frequency')}
                        component={InputRange} min={2} max={31} step={1}
                        onChange={(value: number) => setFieldValue('watering_frequency', value)}
                      />
                    </FormCreate.Control>
                  )
                }
                <FormCreate.Control>
                  <FormCreate.Infos>
                    <Trans i18nKey="plant_form.form_notice">
                      Theses data will be added to our database in order to fill it and let you have a better
                      experience with our app. If you want to participate, let us know by <strong>contacting
                      us</strong>!
                    </Trans>
                  </FormCreate.Infos>
                  <FormCreate.ButtonWrapper>
                    <Button variant="primary" type="submit" data-testid="submit" disabled={isSubmitting}>
                      {submitLabel}
                    </Button>
                    <Button variant="danger" type="reset" disabled={isSubmitting}>
                      {t('plant_form.button.cancel')}
                    </Button>
                  </FormCreate.ButtonWrapper>
                </FormCreate.Control>
              </div>
            </FormCreate.Wrapper>
          </FormCreate>
        );
      }}
    />
  );
};

export default PlantForm;
