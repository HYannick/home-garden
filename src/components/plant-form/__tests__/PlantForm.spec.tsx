import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup } from '@testing-library/react';
import { renderWithTheme } from '../../../theme-wrapper';
import PlantForm from '../PlantForm';

afterEach(cleanup);
describe('PlanForm', () => {
  const defaultProps = {
    t: jest.fn(),
    initialValues: {},
    onSubmit: jest.fn(),
    submitLabel: 'Create',
  };

  it('should render properly', () => {
    const { container } = renderWithTheme(
      <PlantForm {...defaultProps} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the right submitLabel', () => {
    const { getByTestId } = renderWithTheme(
      <PlantForm {...defaultProps} />,
    );
    expect(getByTestId('submit')).toHaveTextContent('Create');
  });

  it('should display the has_moisture_sensor input', () => {
    const props = {
      ...defaultProps,
      initialValues: {
        has_moisture_sensor: true
      }
    };
    const { getByTestId } = renderWithTheme(
      <PlantForm {...props} />,
    );
    expect(getByTestId('moisture-sensor-input')).toBeTruthy();
  });

  it('should display the watering frequency input range', () => {
    const props = {
      ...defaultProps,
      initialValues: {
        has_moisture_sensor: false
      }
    };
    const { getByTestId } = renderWithTheme(
      <PlantForm {...props} />,
    );
    expect(getByTestId('watering-frequency-input')).toBeTruthy();
  });
});
