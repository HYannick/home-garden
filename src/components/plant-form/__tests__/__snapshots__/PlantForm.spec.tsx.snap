// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanForm should render properly 1`] = `
<div>
  <form
    class="css-16qml6p"
  >
    <div
      class="css-8djnc7-ImageUpload"
    >
      <label
        class="css-j7s6en"
        for="picture"
      >
        <div
          class="css-y3dm6m"
        >
          <svg
            viewBox="0 0 515.667 515.667"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M446.533 119.425l-54.117-.283-51.85-77.35c-3.117-4.817-8.5-7.933-14.167-7.933H186.717c-5.667 0-10.767 3.117-13.883 7.65l-53.55 77.633-50.15.283C31.167 119.425 0 150.592 0 188.558v223.55c0 37.967 31.167 69.7 69.133 69.7h377.4c37.967 0 69.133-31.733 69.133-69.7v-223.55c.001-37.966-31.166-69.133-69.133-69.133zm35.134 292.683c0 19.267-15.867 35.7-35.133 35.7h-377.4c-19.267 0-35.133-16.433-35.133-35.7v-223.55c0-19.267 15.867-35.133 35.133-35.133h59.217c5.667 0 10.767-3.117 13.883-7.65l53.55-77.633h121.55l51.567 77.067c3.117 4.533 8.5 7.65 13.883 7.65l63.467.567c19.267 0 35.133 15.867 35.133 35.133v223.55h.283z"
            />
            <path
              d="M259.25 160.225c-66.583 0-120.417 54.117-120.417 120.417S192.95 401.058 259.25 401.058s120.417-54.117 120.417-120.417S325.55 160.225 259.25 160.225zm0 206.833c-47.6 0-86.417-38.817-86.417-86.417s38.817-86.417 86.417-86.417 86.417 38.817 86.417 86.417-38.817 86.417-86.417 86.417z"
            />
            <circle
              cx="420.467"
              cy="210.092"
              r="6.517"
            />
            <path
              d="M420.467 186.575c-13.033 0-23.517 10.483-23.517 23.517s10.483 23.517 23.517 23.517 23.517-10.483 23.517-23.517-10.484-23.517-23.517-23.517zm0 34c-5.95 0-10.483-4.817-10.483-10.483s4.817-10.483 10.483-10.483c5.667 0 10.483 4.817 10.483 10.483s-4.533 10.483-10.483 10.483z"
            />
          </svg>
          <span>
            Provide a photo here
          </span>
        </div>
      </label>
      <input
        class="css-17ejumq"
        id="picture"
        type="file"
      />
    </div>
    <div
      class="css-ov1ktg"
    >
      <div
        class="css-u7rvqb"
      >
        <p>
          Plant infos
        </p>
      </div>
      <div
        class="css-11493n4"
      >
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="custom_name"
          >
            plant_form.field_name
          </label>
          <input
            class="css-1qxlqm1"
            id="custom_name"
            name="custom_name"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="last_watering_date"
          >
            plant_form.field_last_watering_date
          </label>
          <div
            class="DayPicker css-1bwhhw4"
            lang="en"
          >
            <div
              class="DayPicker-wrapper"
              tabindex="0"
            >
              <div
                class="DayPicker-NavBar"
              >
                <span
                  aria-label="Previous Month"
                  class="DayPicker-NavButton DayPicker-NavButton--prev"
                  role="button"
                  tabindex="0"
                />
                <span
                  aria-label="Next Month"
                  class="DayPicker-NavButton DayPicker-NavButton--next"
                  role="button"
                  tabindex="0"
                />
              </div>
              <div
                class="DayPicker-Months"
              >
                <div
                  class="DayPicker-Month"
                  role="grid"
                >
                  <div
                    class="DayPicker-Caption"
                    role="heading"
                  >
                    <div>
                      September 2019
                    </div>
                  </div>
                  <div
                    class="DayPicker-Weekdays"
                    role="rowgroup"
                  >
                    <div
                      class="DayPicker-WeekdaysRow"
                      role="row"
                    >
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Sunday"
                        >
                          Su
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Monday"
                        >
                          Mo
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Tuesday"
                        >
                          Tu
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Wednesday"
                        >
                          We
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Thursday"
                        >
                          Th
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Friday"
                        >
                          Fr
                        </abbr>
                      </div>
                      <div
                        class="DayPicker-Weekday"
                        role="columnheader"
                      >
                        <abbr
                          title="Saturday"
                        >
                          Sa
                        </abbr>
                      </div>
                    </div>
                  </div>
                  <div
                    class="DayPicker-Body"
                    role="rowgroup"
                  >
                    <div
                      class="DayPicker-Week"
                      role="row"
                    >
                      <div
                        aria-disabled="false"
                        aria-label="Sun Sep 01 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="0"
                      >
                        1
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Mon Sep 02 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="-1"
                      >
                        2
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Tue Sep 03 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="-1"
                      >
                        3
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Wed Sep 04 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="-1"
                      >
                        4
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Thu Sep 05 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="-1"
                      >
                        5
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Fri Sep 06 2019"
                        aria-selected="false"
                        class="DayPicker-Day"
                        role="gridcell"
                        tabindex="-1"
                      >
                        6
                      </div>
                      <div
                        aria-disabled="false"
                        aria-label="Sat Sep 07 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--today"
                        role="gridcell"
                        tabindex="-1"
                      >
                        7
                      </div>
                    </div>
                    <div
                      class="DayPicker-Week"
                      role="row"
                    >
                      <div
                        aria-disabled="true"
                        aria-label="Sun Sep 08 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        8
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Mon Sep 09 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        9
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Tue Sep 10 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        10
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Wed Sep 11 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        11
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Thu Sep 12 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        12
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Fri Sep 13 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        13
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Sat Sep 14 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        14
                      </div>
                    </div>
                    <div
                      class="DayPicker-Week"
                      role="row"
                    >
                      <div
                        aria-disabled="true"
                        aria-label="Sun Sep 15 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        15
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Mon Sep 16 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        16
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Tue Sep 17 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        17
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Wed Sep 18 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        18
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Thu Sep 19 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        19
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Fri Sep 20 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        20
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Sat Sep 21 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        21
                      </div>
                    </div>
                    <div
                      class="DayPicker-Week"
                      role="row"
                    >
                      <div
                        aria-disabled="true"
                        aria-label="Sun Sep 22 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        22
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Mon Sep 23 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        23
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Tue Sep 24 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        24
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Wed Sep 25 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        25
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Thu Sep 26 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        26
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Fri Sep 27 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        27
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Sat Sep 28 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        28
                      </div>
                    </div>
                    <div
                      class="DayPicker-Week"
                      role="row"
                    >
                      <div
                        aria-disabled="true"
                        aria-label="Sun Sep 29 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        29
                      </div>
                      <div
                        aria-disabled="true"
                        aria-label="Mon Sep 30 2019"
                        aria-selected="false"
                        class="DayPicker-Day DayPicker-Day--disabled"
                        role="gridcell"
                        tabindex="-1"
                      >
                        30
                      </div>
                      <div
                        aria-disabled="true"
                        class="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
                      />
                      <div
                        aria-disabled="true"
                        class="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
                      />
                      <div
                        aria-disabled="true"
                        class="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
                      />
                      <div
                        aria-disabled="true"
                        class="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
                      />
                      <div
                        aria-disabled="true"
                        class="DayPicker-Day DayPicker-Day--disabled DayPicker-Day--outside"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="has_moisture_sensor"
          >
            plant_form.field_has_moisture_sensor
          </label>
          <label
            class="css-qwug7p"
            for="has_moisture_sensor"
          >
            <input
              type="checkbox"
            />
            <input
              class="css-x6ytko"
              id="has_moisture_sensor"
              name="has_moisture_sensor"
              type="checkbox"
            />
            <span
              class="slider round css-843eyc"
            />
          </label>
        </div>
        <div
          class="css-1diq458"
          data-testid="watering-frequency-input"
        >
          <label
            class="css-he3bhv"
            for="watering_frequency"
          >
            What will your 
            <strong>
              watering frequency
            </strong>
             be?
          </label>
          <div
            class="css-18uljbk"
          >
            <div
              class="css-98r97g"
            >
              <div
                class="css-1wypj0"
                max="31"
                min="2"
                style="cursor: pointer;"
                values="2"
              >
                <div
                  aria-valuemax="31"
                  aria-valuemin="2"
                  aria-valuenow="2"
                  class="css-1r0bhv1"
                  draggable="false"
                  role="slider"
                  style="position: absolute; z-index: 0; cursor: grab; user-select: none; transform: translate(NaNpx, NaNpx);"
                  tabindex="0"
                >
                  2
                </div>
              </div>
            </div>
            <div
              class="css-yvyril"
            >
              <span>
                2
              </span>
              <span>
                31
              </span>
            </div>
            <span
              class="css-1ia2zn2"
            >
              in days
            </span>
          </div>
        </div>
        <div
          class="css-1diq458"
        >
          <p
            class="css-g9ifac"
          >
            Theses data will be added to our database in order to fill it and let you have a better experience with our app. If you want to participate, let us know by 
            <strong>
              contacting us
            </strong>
            !
          </p>
          <div
            class="css-115f2m6"
          >
            <button
              class="css-139qh1f"
              data-testid="submit"
              type="submit"
            >
              Create
            </button>
            <button
              class="css-1r4xqxr"
              type="reset"
            >
              plant_form.button.cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
